File: fabric/js/fabric-core/src/signing.ts
/**
 * Optional BIP-340 Schnorr signing with the Spaces signed-message prefix.
 *
 * Requires the `@noble/curves` package:
 *   npm install @noble/curves
 */

import { schnorr } from "@noble/curves/secp256k1";
import { sha256 } from "@noble/hashes/sha256";

const SPACES_SIGNED_MSG_PREFIX = new TextEncoder().encode(
  "\x17Spaces Signed Message:\n",
);

function hashSignable(msg: Uint8Array): Uint8Array {
  const combined = new Uint8Array(
    SPACES_SIGNED_MSG_PREFIX.length + msg.length,
  );
  combined.set(SPACES_SIGNED_MSG_PREFIX);
  combined.set(msg, SPACES_SIGNED_MSG_PREFIX.length);
  return sha256(combined);
}

/**
 * Sign a message using BIP-340 Schnorr with the Spaces signed-message prefix.
 *
 * Takes raw message bytes (e.g. `recordSet.toBytes()`) and a 32-byte secret key.
 * Returns a 64-byte signature.
 */
export function signMessage(
  msg: Uint8Array,
  secretKey: Uint8Array,
): Uint8Array {
  const hash = hashSignable(msg);
  return schnorr.sign(hash, secretKey);
}

/**
 * Verify a BIP-340 Schnorr signature over a message with the Spaces signed-message prefix.
 */
export function verifyMessage(
  msg: Uint8Array,
  signature: Uint8Array,
  pubkey: Uint8Array,
): boolean {
  const hash = hashSignable(msg);
  return schnorr.verify(signature, hash, pubkey);
}

File: fabric/kotlin/src/main/kotlin/org/spacesprotocol/fabric/Signing.kt
package org.spacesprotocol.fabric

import fr.acinq.secp256k1.Secp256k1
import java.security.MessageDigest
import java.security.SecureRandom

private val SPACES_SIGNED_MSG_PREFIX = "\u0017Spaces Signed Message:\n".toByteArray()

private fun hashSignable(msg: ByteArray): ByteArray {
    val digest = MessageDigest.getInstance("SHA-256")
    digest.update(SPACES_SIGNED_MSG_PREFIX)
    digest.update(msg)
    return digest.digest()
}

/**
 * Sign a message using BIP-340 Schnorr with the Spaces signed-message prefix.
 *
 * Takes raw message bytes (e.g. `recordSet.toBytes()`) and a 32-byte secret key.
 * Returns a 64-byte signature.
 */
fun signMessage(msg: ByteArray, secretKey: ByteArray): ByteArray {
    require(secretKey.size == 32) { "secret key must be 32 bytes, got ${secretKey.size}" }
    val hash = hashSignable(msg)
    val auxRand = ByteArray(32).also { SecureRandom().nextBytes(it) }
    return Secp256k1.signSchnorr(hash, secretKey, auxRand)
}

/**
 * Verify a BIP-340 Schnorr signature over a message with the Spaces signed-message prefix.
 */
fun verifyMessage(msg: ByteArray, signature: ByteArray, pubkey: ByteArray): Boolean {
    require(signature.size == 64) { "signature must be 64 bytes, got ${signature.size}" }
    require(pubkey.size == 32) { "pubkey must be 32 bytes (x-only), got ${pubkey.size}" }
    val hash = hashSignable(msg)
    return Secp256k1.verifySchnorr(signature, hash, pubkey)
}
